packages/runtime/src/runtime.ts

```diff
@@ -459,23 +479,56 @@ export class Runtime {
   }
 
   /**
-   * creates new application and returns application id
-   * @param flags SSCDeployment flags
-   * @param payFlags Transaction parameters
+   * create new application and returns application id
+   * @deprecated `deployApp` should be used instead.
    * @param approvalProgram application approval program
    * @param clearProgram application clear program
+   * @param flags SSCDeployment flags
+   * @param payFlags Transaction parameters
    * @param debugStack: if passed then TEAL Stack is logged to console after
    * each opcode execution (upto depth = debugStack)
    * NOTE - approval and clear program must be the TEAL code as string (not compiled code)
    */
   addApp (
-    flags: AppDeploymentFlags, payFlags: types.TxParams,
     approvalProgram: string, clearProgram: string,
+    flags: AppDeploymentFlags, payFlags: types.TxParams,
     debugStack?: number
   ): DeployedAppTxReceipt {
     this.addCtxAppCreateTxn(flags, payFlags);
     this.ctx.debugStack = debugStack;
-    const txReceipt = this.ctx.addApp(flags.sender.addr, flags, approvalProgram, clearProgram, 0);
+    const txReceipt = this.ctx.deployApp(flags.sender.addr, flags, approvalProgram, clearProgram, 0);
+
+    this.store = this.ctx.state;
+    return txReceipt;
+  }
+
+  /**
+   * deploy a new application and returns application id
+   * @param approvalProgramFileName application approval program filename
+   * @param clearProgramFileName application clear program filename
+   * @param flags SSCDeployment flags
+   * @param payFlags Transaction parameters
+   * @param debugStack: if passed then TEAL Stack is logged to console after
+   * each opcode execution (upto depth = debugStack)
+   * NOTE - approval and clear program must be the TEAL code as string (not compiled code)
+   */
+  deployApp (
+    approvalProgramFileName: string,
+    clearProgramFileName: string,
+    flags: AppDeploymentFlags,
+    payFlags: types.TxParams,
+    scTmplParams?: SCParams,
+    debugStack?: number
+  ): DeployedAppTxReceipt {
+    this.addCtxAppCreateTxn(flags, payFlags);
+    this.ctx.debugStack = debugStack;
+    const approvalProgram = getProgram(approvalProgramFileName, scTmplParams);
+
+    // if clearProgramFileName is empty => clearProgram = ""
+    // else load it by getProgram
+    const clearProgram = clearProgramFileName ? getProgram(clearProgramFileName, scTmplParams) : "";
```

Comment:
a) logic should be handled for both - approval & clear programs
b) i think we simply need to check if it's a fileName first, if not then proceed with `getProgram`. something like:
 ```js
  const clearProgram = clearProgramFileName.endsWith(".teal" OR ".py") ? getProgram(clearProgramFileName, scTmplParams) : clearProgramFileName;
   ```

Reply:
BDW, i think we should the compilation in `ctx.deployApp` (so move this logic to `ctx.ts`), this way we will also handle `runtime.getAppInfoFromName`. Otherwise it will always store TEAL code as key even if user passes file name.

And probably which is why you take both filenames & tealcode in examples
```js
  const approvalProgramFileName = 'crowdFundApproval.teal';
  const clearProgramFileName = 'crowdFundClear.teal';
  const approvalProgram = getProgram(approvalProgramFileName);
  const clearProgram = getProgram(clearProgramFileName);
```

Reply:
In runtime.updateApp also pass program instead of name file, so why i wrote code like above.

Reply:
i updated it